File: 107. Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // stack: O(N)time, O(depth)space, 33.26%, ok
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ret;
        if(!root) return ret;
        queue<TreeNode*> q;
        q.push(root);
        stack<vector<int>> filo;
        while(q.size()){
            int size = q.size();
            vector<int> level;
            while(size--){
                TreeNode* temp = q.front();
                q.pop();
                level.push_back(temp->val);
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
            filo.push(level);
        }
        while(filo.size()){
            ret.push_back(filo.top());
            filo.pop();
        }
        return ret;
    }

    // STL: O(N)time, O(1)space, 33.26%, ok
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ret;
        if(!root) return ret;
        queue<TreeNode*> q;
        q.push(root);
        while(q.size()){
            int size = q.size();
            vector<int> level;
            while(size--){
                TreeNode* temp = q.front();
                q.pop();
                level.push_back(temp->val);
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
            ret.push_back(level);
        }
        reverse(ret.begin(), ret.end());
        return ret;
    }
};

File: 107. Binary Tree Level Order Traversal II.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # bfs, queue: O(N) time, O(1) space
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        queue, ret = deque([root]), []
        while queue:
            level_size = len(queue)
            ret.append([x.val for x in queue])
            while level_size:
                level_size -= 1
                node = queue.popleft()
                queue += filter(None, [node.left, node.right])
        return ret[::-1]
            
        